### PR TITLE
Feature/codegenscopetracking

### PR DESCRIPTION
**Pull request by mirakst**
Code generator now tracks scope.
Arrays are now malloced, and freed at end of scope.
These changes were necessary to allow functions returning arrays.

### CODE REVIEW

File: STEP/CodeGeneration/CodeGenerationVisitor.cs [hunk 1]
```diff
@@ -32,27 +32,46 @@ private void EmitAppend(string line)
...
+                EmitAppend("boolean" + suffix);
                 break;
             case TypeVal.Blank:
-                EmitAppend("void ");
+                EmitAppend("void" + suffix);
                 break;
         }
     }
+
+    private void EnterScope() {
```
Comment (damniko) [inline STEP/CodeGeneration/CodeGenerationVisitor.cs:57]: Der skal helt sikkert noget dokumentation til det her - think of the children!

File: STEP/CodeGeneration/CodeGenerationVisitor.cs [hunk 2]
```diff
@@ -144,29 +163,33 @@ public void Visit(BoolNode n)
...
+        // if (n.IdRight is not null)
+        // {
+        //     EmitAppend(" = ");
+        //     n.IdRight.Accept(this);
+        // }
+        // else if (n.Right is not null)
+        // {
+        //     EmitAppend(" = ");
+        //     n.Right.Accept(this);
+        // }
```
Comment (damniko) [inline STEP/CodeGeneration/CodeGenerationVisitor.cs]: Skal det her vækski?

**Review by damniko — APPROVED** [review]
very smarts!

File: STEP/CodeGeneration/CodeGenerationVisitor.cs (hunk 2 above)
Comment (mirakst) [reply STEP/CodeGeneration/CodeGenerationVisitor.cs]: yes!

**Review by Sarmisuper — APPROVED** [review]
lgtm

File: STEP/CodeGeneration/CodeGenerationVisitor.cs (hunk 1 above)
Comment (mirakst) [reply STEP/CodeGeneration/CodeGenerationVisitor.cs:57]: I am
EliteSquadRichGIF

**Comment by mirakst** [comment]
[Always clear arrays when variable goes out of scope](https://trello.com/c/aY2q8Nh3/107-always-clear-arrays-when-variable-goes-out-of-scope)

**Pull request merged**